### DISCUSSION TITLE
Unexpected "Invalid literal value" error in z.union()

### DISCUSSION THREAD

**Discussion by tomohirohiratsuka** [colinhacks/zod#2067] [Q&A] [answered]
Hi, First of all, thank you for creating very useful validation library.
I've faced wired(?) behavior while using z.union().
Is this bug? or if this is expected behavior how can I avoid it to show wrong errors?

# Problem
zod shows invalid literal value though it has correct value.

# Reproduction Repo
I've made a simple reproduction repo [here](https://github.com/tomohirohiratsuka/zod-union-debug)

# Reproduction Step
In the reproduction repo above...

```terminal
npm i 
```

```terminal
npm run dev
```
 1. Open [http://localhost:3000](http://localhost:3000).
 2. Click submit.
 3. It shows Invalid URL error as expected.
 4. Select "Designer" in Job select box.
 5. It shows, Invalid literal value, expected "engineer"

<img width="288" alt="スクリーンショット 2023-02-19 10 30 10" src="https://user-images.githubusercontent.com/32760882/219907250-7bd9169f-8173-40be-ba15-73adc61d1afc.png">
<img width="229" alt="スクリーンショット 2023-02-19 10 30 26" src="https://user-images.githubusercontent.com/32760882/219907257-d5a4b092-6095-4643-aea1-0cce1f11e4d0.png">
<img width="345" alt="スクリーンショット 2023-02-19 10 30 44" src="https://user-images.githubusercontent.com/32760882/219907260-40430039-a70b-4529-9173-ffda8fbf7523.png">

# Schema
```typescript
const developmentLanguageSchema = z.union([z.literal('typescript'), z.literal('go'), z.literal('rust')])
const toolSchema = z.union([z.literal('Adobe XD'), z.literal('Figma'), z.literal('Sketch')])
const schema = z.union([
    z.object({
        job: z.literal('engineer'),
        language: developmentLanguageSchema,
        graduate: z.literal('bachelor'),
    }),
    z.object({
        job: z.literal('engineer'),
        language: developmentLanguageSchema,
        graduate: z.literal('master'),
        url: z.string().url()
    }),
    z.object({
        job: z.literal('designer'),
        tool: toolSchema,
        graduate: z.literal('bachelor'),
    }),
    z.object({
        job: z.literal('designer'),
        tool: toolSchema,
        graduate: z.literal('master'),
        url: z.string().url()
    })
])
```

**Comment by JacobWeisenburger** [comment, marked as answer]
There are a few things happening here:

In zod, if you try to validate an object that has an error on one of it's fields, it will consider the whole object invalid. Since in your example, you don't have any value in `url`, it considers the whole object invalid, even though `job`, `tool`, and `graduate` are valid.

I'm not super familiar with `react-hook-form`, but I'm guessing that it only shows errors on `touched` or `dirty` fields. So when you load the form initially, `job` is `engineer`, which technically zod considers invalid because the you are missing `url`. But since `job` is not `touched` nor `dirty`, `react-hook-form` doesn't show the error. Once you change `job` to `designer`, then that field is `touched` and `dirty`, so `react-hook-form` shows the error. Once all fields pass validation, then the form will be valid and the errors should go away.

Let me know if you have any questions about this. I'm happy to help.

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger
